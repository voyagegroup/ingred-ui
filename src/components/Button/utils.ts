import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
