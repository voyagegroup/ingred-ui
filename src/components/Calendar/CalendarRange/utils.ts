import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
