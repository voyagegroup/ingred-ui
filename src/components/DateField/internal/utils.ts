export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
