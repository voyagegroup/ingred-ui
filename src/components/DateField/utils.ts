import { useMemo } from "react";

type Sections = {
  start: number;
  end: number;
  value: string;
  editable: boolean;
};

/**
 * 何らかのフォーマットで入ってくる日付を開始位置と終了位置と値を持つセクションに分割する
 * useDateField で format された日付操作を汎用的に行うために必要なプロパティを返す
 * 例) 2023-01-02 -> [
 *   { start: 0, end: 3, value: "2023", editable: true },
 *   { start: 4, end: 4, value: "-", editable: false },
 *   { start: 5, end: 6, value: "01", editable: true },
 *   { start: 7, end: 7, value: "-", editable: false },
 *   { start: 8, end: 9, value: "02", editable: true }
 * ]
 *
 * @param formattedDate 何らかのフォーマットで入ってくる日付
 * @returns 開始位置と終了位置と値を持つセクション
 */
export const getSections = (formattedDate?: string | null) => {
  if (!formattedDate) {
    return [];
  }

  const sections: Sections[] = [];
  let start = 0;
  let isPrevCharDigit = !isNaN(Number(formattedDate[0]));

  for (let index = 1; index <= formattedDate.length; index++) {
    const currentChar = formattedDate[index];
    const isCurrentCharDigit =
      !isNaN(Number(currentChar)) && currentChar !== " ";

    if (
      isCurrentCharDigit !== isPrevCharDigit ||
      currentChar === " " ||
      index === formattedDate.length
    ) {
      sections.push({
        start,
        end: index - 1,
        value: formattedDate.slice(start, index),
        editable: isPrevCharDigit,
      });
      start = index;
      isPrevCharDigit = isCurrentCharDigit;
    }
  }

  return sections;
};

/**
 * 開始位置と終了位置と値を持つセクションをフォーマットされた日付に変換する
 */
export const formatString = (sectionsWithCharactor: Sections[]) =>
  sectionsWithCharactor.map((section) => section.value).join("");

type ReactRef<T> =
  | React.RefCallback<T>
  | React.MutableRefObject<T>
  | React.ForwardedRef<T>
  | string
  | null
  | undefined;

// from: https://github.com/voyagegroup/ingred-ui/blob/master/src/hooks/useMergeRefs.ts
export function useMergeRefs<T>(...refs: ReactRef<T>[]): React.Ref<T> {
  return useMemo(() => {
    if (refs.every((ref) => ref === null)) {
      return null;
    }
    return (refValue: T) => {
      for (const ref of refs) {
        if (typeof ref === "function") {
          ref(refValue);
        } else if (ref && typeof ref !== "string") {
          ref.current = refValue;
        }
      }
    };
  }, [refs]);
}
