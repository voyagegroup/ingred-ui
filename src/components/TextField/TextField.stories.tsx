import * as React from "react";
import { Story } from "@storybook/react/types-6-0";
import TextField, { TextFieldProps } from "./TextField";
import { Flex, Spacer, Typography } from "..";

export default {
  title: "Components/Inputs/TextField",
  component: TextField,
  parameters: {
    docs: {
      description: {
        component: `
The wrapper of ${"`<Input />`"} component.
`,
      },
    },
  },
};

export const Example: Story<TextFieldProps> = (args) => <TextField {...args} />;

Example.args = {
  placeholder: "search",
  icon: "search",
};

export const Password = () => <TextField defaultValue="hoge" type="password" />;

export const Error = () => (
  <Flex display="flex">
    <div>
      <Typography weight="bold">Input</Typography>
      <Spacer pt={2} />
      <TextField type="number" errorText="Please input number only" />
    </div>
    <Spacer pl={5} />
    <div>
      <Typography weight="bold">Text field</Typography>
      <Spacer pt={2} />
      <TextField errorText="Invalid input" multiline={true} />
    </div>
  </Flex>
);
