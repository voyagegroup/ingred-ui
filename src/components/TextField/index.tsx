export { default, TextFieldProps } from "./TextField";
