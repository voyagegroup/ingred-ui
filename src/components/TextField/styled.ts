import styled from "styled-components";

export const Container = styled.div``;

export const InputContainer = styled.div<{
  hasLeftIcon: boolean;
  hasRightIcon: boolean;
}>`
  position: relative;
  display: flex;
  align-items: center;
  width: 100%;
  & > input {
    display: inline-flex;
    flex: 1;
    ${({ hasRightIcon }) => (hasRightIcon ? "padding-right: 32px" : "")};
    ${({ hasLeftIcon }) => (hasLeftIcon ? "padding-left: 35px" : "")};
  }
`;

const IconContainer = styled.div`
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  cursor: pointer;
`;

export const LeftIconContainer = styled(IconContainer)`
  left: 10px;
`;

export const RightIconContainer = styled(IconContainer)`
  right: 10px;
`;
