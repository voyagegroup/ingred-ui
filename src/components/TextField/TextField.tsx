import * as React from "react";
import * as Styled from "./styled";
import Input from "../Input";
import { InputProps } from "../Input/Input";
import ErrorText from "../ErrorText";
import Spacer from "../Spacer";
import Icon from "../Icon";
import { useTheme } from "../../themes";
import { IconName } from "../Icon/Icon";

export type TextFieldProps = InputProps & {
  errorText?: string;
  icon?: IconName;
  type?: string;
  inputRef?: React.Ref<HTMLInputElement | HTMLTextAreaElement>;
};

const TextField = React.forwardRef<HTMLDivElement, TextFieldProps>(
  function TextField({ errorText, icon, type, inputRef, ...rest }, ref) {
    const [show, setShow] = React.useState(false);
    const theme = useTheme();
    const handleToggleShowPassword = () => {
      setShow(!show);
    };

    return (
      <Styled.Container ref={ref}>
        <Styled.InputContainer
          hasLeftIcon={icon != null}
          hasRightIcon={type === "password"}
        >
          {icon != null && (
            <Styled.LeftIconContainer onClick={handleToggleShowPassword}>
              <Icon name={icon} size="md" color={theme.palette.gray.dark} />
            </Styled.LeftIconContainer>
          )}
          <Input
            ref={inputRef}
            {...rest}
            error={!!errorText}
            type={show ? "text" : type}
          />
          {type === "password" && (
            <Styled.RightIconContainer onClick={handleToggleShowPassword}>
              <Icon
                name={show ? "eye" : "eye_off"}
                size="md"
                color={theme.palette.black}
              />
            </Styled.RightIconContainer>
          )}
        </Styled.InputContainer>
        {errorText && (
          <Spacer pt={1}>
            <ErrorText>{errorText}</ErrorText>
          </Spacer>
        )}
      </Styled.Container>
    );
  },
);

export default TextField;
