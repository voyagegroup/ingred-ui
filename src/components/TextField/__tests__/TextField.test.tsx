import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import TextField from "../";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("TextField component testing", () => {
  afterEach(cleanup);

  test("TextField", () => {
    const { asFragment } = renderWithThemeProvider(<TextField />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("TextField with icon", () => {
    const { asFragment } = renderWithThemeProvider(<TextField icon="search" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("TextField passward", () => {
    const { asFragment } = renderWithThemeProvider(
      <TextField type="password" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
