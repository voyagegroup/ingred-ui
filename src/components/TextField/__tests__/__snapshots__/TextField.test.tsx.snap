// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField component testing TextField 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL"
  >
    <div
      class="sc-gsDKAQ iXEfkO"
    >
      <input
        class="sc-jRQBWg bfvJhB"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`TextField component testing TextField passward 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL"
  >
    <div
      class="sc-gsDKAQ kJFxMN"
    >
      <input
        class="sc-jRQBWg bfvJhB"
        type="password"
      />
      <div
        class="sc-dkPtRN sc-eCImPb fVDQHx vbNRz"
      >
        <div
          class="sc-gKclnd edtajv"
          size="18"
        >
          <svg
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0,0H24V24H0Z"
              fill="none"
            />
            <path
              d="M17.882,19.3A11,11,0,0,1,1.181,12a10.982,10.982,0,0,1,3.34-6.066L1.392,2.808,2.807,1.393l19.8,19.8-1.415,1.414-3.31-3.31ZM5.935,7.35A8.965,8.965,0,0,0,3.223,12a9,9,0,0,0,13.2,5.838L14.4,15.81A4.5,4.5,0,0,1,8.19,9.6Zm6.979,6.978L9.672,11.086a2.5,2.5,0,0,0,3.241,3.241Zm7.893,2.264-1.431-1.43A8.935,8.935,0,0,0,20.777,12,9.005,9.005,0,0,0,9.552,5.338L7.974,3.76A11.01,11.01,0,0,1,22.819,12a10.947,10.947,0,0,1-2.012,4.592ZM11.723,7.508a4.5,4.5,0,0,1,4.769,4.769l-4.77-4.769Z"
              fill="#041C33"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TextField component testing TextField with icon 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL"
  >
    <div
      class="sc-gsDKAQ fomYWt"
    >
      <div
        class="sc-dkPtRN sc-hKwDye fVDQHx bmmLIk"
      >
        <div
          class="sc-gKclnd edtajv"
          size="18"
        >
          <svg
            viewBox="0 0 18 18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h18v18H0z"
              fill="none"
            />
            <path
              d="M14.023,12.963l3.212,3.211-1.061,1.061-3.211-3.212a6.751,6.751,0,1,1,1.061-1.06Zm-1.5-.556a5.248,5.248,0,1,0-.112.112l.112-.112Z"
              fill="#959FA9"
              transform="translate(-0.5 -0.5)"
            />
          </svg>
        </div>
      </div>
      <input
        class="sc-jRQBWg bfvJhB"
      />
    </div>
  </div>
</DocumentFragment>
`;
